### ISSUE TITLE
22.1.4_1: Creating new VLAN interfaces fails to bring them up

### ISSUE THREAD

**Issue by matsimon** [opnsense/core#5650]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

Creating a new VLAN after the upgrade (and relevant changes in VLAN naming / structuring) with 22.1.4_1 creating newA clear and concise description of what the bug is, including last known working version (if any).

*Tip*: to validate your setup was working with the previous version, use opnsense-revert (https://docs.opnsense.org/manual/opnsense_tools.html#opnsense-revert)

**To Reproduce**

1. Go to Interfaces > Other Types > VLAN, Click on the "+"
2. Select a parent interface (that already has other VLANs in my case), assign a VLAN ID, click on save
3. Go to Interfaces > Assignments, click on the "+"
4. Select the newly-created VLAN interface from the dropdown list, then click on Save
5. Go to Interfaces > (Name of new OPT interface): Select "Enable interface", define i.e. an IPv4 address
6. Click on "Save", then apply the changes what should then enable and bring up the new VLAN interface

The issue is definitely reproducible for what I can tell after trying with 3 VLAN interface definitions.

**Expected behavior**

A VLAN interface is created and switches up (as the parent interface is also up) like previously-created ones, instead the interface remains down in the Web UI.

I've also checked the output of 'ifconfig' to realize now vlanXY interface was defined, then saw an error in /var/log/system as mentioned in "relevant log files".

**Describe alternatives you considered**

By replacing the name with the old way (i.e. re1_vlan112) in config.xml, then triggering a reload by changing the IP, then applying the settings in the UI, the VLAN interface came up working. Also: VLAN interfaces created before the upgrade also work flawlessly.

Defintely not "nice", but it did work at least in my case here as a workaround.

**Screenshots**

(Not availbale.)

**Relevant log files**

The following message appears in /var/log/system/(datecode).log
```
<11>1 2022-03-27Txx:xx:xx+00:00 apu1.removed.example opnsense 79614 - [meta sequenceId="9"] /interfaces.php: Unable to configure non-existent interface opt6 (vlan02)
```

**Additional context**

I don't have the ability to compare it on another system that easily.

**Environment**

Software version used and hardware type if relevant, e.g.:

OPNsense 21.1.4_1 (amd64, OpenSSL).
PCEngines APU1, 4GB
Network Realtek RTL8111E

**Comment by AdSchellevis** [comment]
> 1. Go to Interfaces > Other Types > VLAN, Click on the "+"

did you hit apply after the adding a new vlan?

**Comment by matsimon** [comment]
Hi and thank you asking for more precise information to reproduce the issue I ran into. Yes indeedt actually I did omit the apply button (didn't realize it to be honest).

Thanks to ZFS BE I temporarily booted back into 21.1.3 to compare the UI behavior: There is no apply button appearing on the preceding version. Yes, I was using the 221.4 UI under the assumption of how 22.1.3 and older releases worked. I did this under the assumption that I could hit apply also only after configuring / enabling the VLAN interface in one go. Sometimes it's is actually helpful to be able to prepare configuration changes in one go and only activate them in a final go only hiting Apply then.

Shouldn't the code check if the VLAN interface is already defined before it attempts enabling and setting IP addresses first? At least that's what I'd assume initially.

**Comment by fichtner** [comment]
Eventually the code should configure the device assigned to the interface. We've only recently made an effort to track these relationships in the code base (22.1.3) for GIF/GRE/bridges, but more to come.

**Comment by fichtner** [comment]
We've added additional validation in 4a754e02bdb for 22.1.5 to make the situation visible (apply missing).

**Comment by matsimon** [comment]
Thank you @fichtner !

**Comment by jtamplin** [comment]
FWIW - this issue just helped me as I was pulling my hair out installing OPNsense 23.7-amd64 on a GoWin R86S.  I did the initial setup, and created VLANs on the 10G port, and I could not ping anything on the other side of them - they would fail with "no route to host".  ifconfig showed the subinterfaces were up, but in the interface assignment page they showed red even though the parent interface was green.

Marking the vlan interfaces as disabled, saving/applying, then marking them as enabled, saving/applying fixed the problem.

This might be related or not, but the workaround certainly helped me.

**Comment by davidsenk** [comment]
I can confirm I experienced the same issue as @jtamplin 

No matter how many times I recreated the new vlans, they would just absolutely refuse to communicate to anything else on the vlan.

I could see the communication working on both the host, and on opnsense with tshark (on host) and tcpdump (on opnsense), so switch was ruled out as the problem. 

I could see arp packets being sent by opnsense to host (Who has? packets), and the host tried to respond to opnsense via the address in the arp discovery packets.

I could see the arp packets being recevied by opnsense on the "master" interface (via tcpdump) but *nothing* got sent back to the vlan interface itself... thus preventing two way communication from working successfully.

I spent an entire day banging my head against the desk trying to get new vlans to work. It was driving me insane. All of the existing vlans worked fine, just new ones I created *absolutely would not work*

Eventually I logged into serial console (I have a Raspberry Pi permanently connected to serial port, so this was being emitted and received by the Pi) for opnsense and... if was filled with this quite odd message:

```
[ 323] freebsd_generic_rx_handler Warning: RX packet intercepted, but no emulated adapter
```

I could recreate that message by trying to "ping" opnsense from *any* new vlans created. Even after clicking apply every step of the way, and enabling / disabling the vlan interface from web ui.

From failing to find anything on Google to get it solved... I decided to try just giving opnsense a reboot...

And... it worked! After reboot all of the "new" vlans I've been banging my head against desk to get working magically started working after the reboot. Communication between hosts and opnsense was working!

Note: I did also see a few of these repeated shortly after boot, but I haven't seen them again... No idea if this is possibly related:

```
[1076] generic_netmap_dtor       Emulated netmap adapter for lagg1 destroyed
[1171] generic_netmap_attach     Emulated adapter for lagg1 created (prev was NULL)
[ 319] generic_netmap_register   Emulated adapter for lagg1 activated
[1171] generic_netmap_attach     Emulated adapter for lagg0 created (prev was NULL)
[1076] generic_netmap_dtor       Emulated netmap adapter for lagg0 destroyed
[ 319] generic_netmap_register   Emulated adapter for lagg0 activated
```

Note: lagg1 is my "LAN" interface with all of the vlans on it. lagg0 is "WAN" interface

And for good measure, here is my NIC information from dmesg:

```
igc0: <Intel(R) Ethernet Controller I226-V> mem 0x80200000-0x802fffff,0x80300000-0x80303fff at device 0.0 on pci2
igc0: Using 1024 TX descriptors and 1024 RX descriptors
igc0: Using 4 RX queues 4 TX queues
igc0: Using MSI-X interrupts with 5 vectors
igc0: Ethernet address: 60:be:b4:0d:62:c0
igc0: netmap queues/slots: TX 4/1024, RX 4/1024
igc1: <Intel(R) Ethernet Controller I226-V> mem 0x80000000-0x800fffff,0x80100000-0x80103fff at device 0.0 on pci3
igc1: Using 1024 TX descriptors and 1024 RX descriptors
igc1: Using 4 RX queues 4 TX queues
igc1: Using MSI-X interrupts with 5 vectors
igc1: Ethernet address: 60:be:b4:0d:62:c1
igc1: netmap queues/slots: TX 4/1024, RX 4/1024
igc2: <Intel(R) Ethernet Controller I226-V> mem 0x7fe00000-0x7fefffff,0x7ff00000-0x7ff03fff at device 0.0 on pci4
igc2: Using 1024 TX descriptors and 1024 RX descriptors
igc2: Using 4 RX queues 4 TX queues
igc2: Using MSI-X interrupts with 5 vectors
igc2: Ethernet address: 60:be:b4:0d:62:c2
igc2: netmap queues/slots: TX 4/1024, RX 4/1024
igc3: <Intel(R) Ethernet Controller I226-V> mem 0x7fc00000-0x7fcfffff,0x7fd00000-0x7fd03fff at device 0.0 on pci5
igc3: Using 1024 TX descriptors and 1024 RX descriptors
igc3: Using 4 RX queues 4 TX queues
igc3: Using MSI-X interrupts with 5 vectors
igc3: Ethernet address: 60:be:b4:0d:62:c3
igc3: netmap queues/slots: TX 4/1024, RX 4/1024
igc0: link state changed to UP
igc1: link state changed to UP
igc2: link state changed to UP
igc3: link state changed to UP
igc0: link state changed to DOWN
lagg0: IPv6 addresses on igc0 have been removed before adding it as a member to prevent IPv6 address scope violation.
igc1: link state changed to DOWN
lagg0: IPv6 addresses on igc1 have been removed before adding it as a member to prevent IPv6 address scope violation.
igc2: link state changed to DOWN
lagg1: IPv6 addresses on igc2 have been removed before adding it as a member to prevent IPv6 address scope violation.
igc3: link state changed to DOWN
lagg1: IPv6 addresses on igc3 have been removed before adding it as a member to prevent IPv6 address scope violation.
igc2: promiscuous mode enabled
igc3: promiscuous mode enabled
igc0: promiscuous mode enabled
igc1: promiscuous mode enabled
igc0: link state changed to UP
igc1: link state changed to UP
igc2: link state changed to UP
igc3: link state changed to UP
igc0: promiscuous mode disabled
igc1: promiscuous mode disabled
igc0: promiscuous mode enabled
igc1: promiscuous mode enabled
```

I've also included the pcap from opnsense showing that the "lagg1" interface does indeed receive the tagged packets correctly, but *somehow* the vlan interface doesn't actually know about them. Again, rebooting opnsense got everything working again, but doesn't explain *why* this happened. Every other vlan I created before yesterday *worked fine* without needing to do anything "magical"

[pcap.zip](https://github.com/user-attachments/files/15942019/pcap.zip)

**Comment by AdSchellevis** [comment]
@davidsenk to debug vlan issues if/when they appear, use `ifconfig` on the machine (or use the interfaces overview page) to compare the ui settings with the ones being configured on the machine itself. Usually nothing magical happens, most common issues are related to settings not being applied (forgotten to press a button). If there's a step by step scenario which results in a missing tag, just open a new ticket, commenting on closed tickets is less useful.

**Comment by davidsenk** [comment]
@AdSchellevis 

I do actually have a copy of the ifconfig output from before I rebooted!

```
vlan0.3901: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        description: vlan3901 (opt9)
        options=4000000<NOMAP>
        ether 60:be:b4:0d:62:c2
        inet6 fe80::62be:b4ff:fe0d:62c2%vlan0.3901 prefixlen 64 scopeid 0x13
        inet6 --PUBLIC ADDRESS OMITTED-- prefixlen 64
        inet 172.16.200.2 netmask 0xffffff00 broadcast 172.16.200.255
        groups: vlan
        vlan: 3901 vlanproto: 802.1q vlanpcp: 0 parent interface: lagg1
        media: Ethernet autoselect
        status: active
        nd6 options=121<PERFORMNUD,AUTO_LINKLOCAL,NO_DAD>
```

I'm not entirely sure I have enough information to be able to reproduce the issue. I *think* it may be uptime dependent. I had 0 issues creating vlans before ~2 days ago. And vlans Ive created after rebooting also work fine, just the two vlans I was trying to create before rebooting (3901, and 1000) could send packets from opnsense, and get received on lagg1 and on the hosts (confirmed with tcpdump and tshark), but opnsense was not delivering it back to the vlan0.3901 / vlan0.1000 interfaces for some reason (as shown in the pcap files).

From comparing the above (pre-reboot) to the current ifconfig output for the vlan... it appears identical to me...

```
vlan0.3901: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        description: vlan3901 (opt9)
        options=4000000<NOMAP>
        ether 60:be:b4:0d:62:c2
        inet6 fe80::62be:b4ff:fe0d:62c2%vlan0.3901 prefixlen 64 scopeid 0xe
        inet6 --PUBLIC ADDRESS OMITTED-- prefixlen 64
        inet 172.16.200.2 netmask 0xffffff00 broadcast 172.16.200.255
        groups: vlan
        vlan: 3901 vlanproto: 802.1q vlanpcp: 0 parent interface: lagg1
        media: Ethernet autoselect
        status: active
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>

```

The only difference I see is the IPv6 link-local `scopeid` is `0x13` pre-reboot and `0xe` post-reboot; as well as the nd6 being 121 and also having NO_DAD prior to reboot. I don't think this would've made a difference as I could not communicate over IPv6 or IPv4, or even at L2, the packets just weren't coming back to the vlan0.3901 interface (as shown in the pcap file), even though lagg1 *did* receive them.

If I can come up with reproducible steps, I'll definitely open a new issue and reference these comments; but for now, this is just me documenting what I currently know to *possibly* help anyone else looking around on Google / etc. If it does end up being uptime dependent, I'll also be able to point back to this ;)

Thanks!
